### PR TITLE
fix: EndDelimiter not working as expected since 4.29

### LINKED ISSUE #6150: EndDelimiter '/' not working as expected since 4.29
### Search first

- [X] I searched and no similar issues were found

### Description

We are deploying SQL and PLSQL scripts on an Oracle Database via Liquibase. We use ‘/’ as endDelimiter because it’s compatible with native SQL and PLSQL.

Since 4.29 this is some kind of broken. Before only a single "/" as first and only character of a new line was recognized as endDelimiter. Now also a "/" at the end of a line is interpreted as endDelimiter. This is not compatible with native SQL anymore.

### Steps To Reproduce

Code similar to the following was leading to an error because of false code splitting

```
-- liquibase formatted sql
-- changeset abc:ABC_PROC runOnChange:true stripComments:false endDelimiter:/
-- comment Initial Creation
CREATE OR REPLACE PROCEDURE ABC_PROC…IS
…
BEGIN
…
i := j /
 2;
…
END;
/
```

### Expected/Desired Behavior

"/" should only interpreted as endDelimiter if it's the only character on a new line.

### Liquibase Version

4.29

### Database Vendor & Version

Oracle 19

### Liquibase Integration

CLI

### Liquibase Extensions

_No response_

### OS and/or Infrastructure Type/Provider

Docker

### Additional Context

_No response_

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR (Thank you!)

### PR DESCRIPTION
**Pull request by filipelautert**
## Description

Consider that you have sql endDelimiter = "/" , and the following SQL:
```
a = var1 / 
var2;
```

How Liquibase should handle it? Should we split or divide? It can be understood as:

```
a = var1 / var2; --  'division behavior'
```
or
```
a = var1 ;   
var2; -- 'strict behavior'
```

Historically, we had:

* Until v4.9.0 - Division behavior
* v4.9.0 (2022.03.17) - changed to Strict behavior - https://github.com/liquibase/liquibase/pull/2135
* v4.15.0 (2022.08.04) - rolled back to Division behavior - https://github.com/liquibase/liquibase/pull/3118
* v4.29.0 (2024-07-28) - changed to Strict behavior again! - https://github.com/liquibase/liquibase/pull/5704

Versions v4.9.0 to v4.14.x (only 5 months) were `strict` while all other versions are `division`. And with 4.29.0 we restored `strict`  because of issue https://github.com/liquibase/liquibase/issues/5359 (this is a case where an upgrade is being done from Liquibase 4.9.1).

As we have 2 different behaviors we decided to support both by using the `strict` flag. So: 

* Division behavior has been suported for longer and it's the default behavior
* Strict behavior is optional and can be enabled by using the `--scrict` parameter ( https://docs.liquibase.com/parameters/strict.html ) .

This is being documented so we don't change it again.

Fixes #6150

### CODE REVIEW

File: liquibase-integration-tests/src/test/java/liquibase/dbtest/oracle/OracleIntegrationTest.java [hunk 1]
```diff
@@ -236,4 +235,23 @@ public void testChangeLogGenerationForTableWithGeneratedColumn() throws Exceptio
         String generatedChangeLog = baos.toString();
         assertTrue("Text '" + textToTest + "' not found in generated change log: " + generatedChangeLog, generatedChangeLog.contains(textToTest));
     }
+
+    @Test
+    public void verifySlashReplacementOnStrictMode() throws DatabaseException {
+        assumeNotNull(getDatabase());
```
Comment (MalloD12) [inline liquibase-integration-tests/src/test/java/liquibase/dbtest/oracle/OracleIntegrationTest.java:241]: We can clean up a bit this test by removing the `clearDatabase()` calls.

File: liquibase-standard/src/main/java/liquibase/util/StringUtil.java [hunk 2]
```diff
@@ -110,8 +110,13 @@
             }
 
             if (isInClause == 0 && splitStatements && (piece instanceof String) && isDelimiter((String) piece, previousPiece, endDelimiter)) {
-                String sentenceWithoutDelimiter = removeEndDelimiterIfItsASlash(endDelimiter, currentString);
-                String trimmedString = sentenceWithoutDelimiter.isEmpty() ? StringUtil.trimToNull(currentString.toString()):StringUtil.trimToNull(sentenceWithoutDelimiter);
+                String trimmedString;
+                if (Boolean.TRUE.equals(GlobalConfiguration.STRICT.getCurrentValue())) {
+                    String sentenceWithoutDelimiter = removeEndDelimiterIfItsASlash(endDelimiter, currentString);
+                    trimmedString = sentenceWithoutDelimiter.isEmpty() ? StringUtil.trimToNull(currentString.toString()) : StringUtil.trimToNull(sentenceWithoutDelimiter);
```
[the remark this answers is not in the document]
Comment (MalloD12) [reply liquibase-standard/src/main/java/liquibase/util/StringUtil.java:116]: We can use `StringUtils` class from `org.apache.commons.lang3`.

File: liquibase-integration-tests/src/test/java/liquibase/dbtest/oracle/OracleIntegrationTest.java [hunk 3]
```diff
@@ -22,6 +23,7 @@
 import liquibase.statement.core.RawParameterizedSqlStatement;
 import liquibase.structure.core.Index;
 import org.junit.Assert;
+import org.junit.Ignore;
```
Comment (rberezen) [inline liquibase-integration-tests/src/test/java/liquibase/dbtest/oracle/OracleIntegrationTest.java:26]: I guess we do not need this

**Comment by rberezen** [comment]
testing blocked by https://datical.atlassian.net/browse/DAT-18440

**Comment by filipelautert** [comment]
Blocked by https://github.com/liquibase/liquibase-pro/pull/1909 . CC @tati-qalified

**Comment by rberezen** [comment]
@tati-qalified @filipelautert this should be unblocked now

**Comment by tati-qalified** [comment]
@rberezen @filipelautert using `--strict=true` doesn't fail anymore!

This changeset generates the same code, no matter if strict is true or false:
```
-- liquibase formatted sql
-- changeset AUTHOR:CHANGESET_NAME endDelimiter:GO rollbackEndDelimiter:GO runAlways:true runOnChange:true dbms:mssql
declare cursorToProcess cursor local
	read_only forward_only fast_forward
	for
		SELECT * FROM DATABASECHANGELOG;

open cursorToProcess;
```

Generates:
```
-- Changeset mssql.sql::CHANGESET_NAME::AUTHOR
declare cursorToProcess cursor local
	read_only forward_only fast_forward
	for
		SELECT * FROM DATABASECHANGELOG;

open cursorToProcess;
GO
```

I think we're good to go with this

**Review by tati-qalified — APPROVED** [review]

**Review by rberezen — APPROVED** [review]

**Comment by rberezen** [comment]
@adrian-velonis1 @AMBERMW13 
we need to update this section:
```
Newline behavior with a / end delimiter
In Liquibase 4.8.0 to 4.14.0, using the delimiter / (forward slash) in a PL/SQL block, for example as a division operator, causes Liquibase to fail on Oracle databases:

Caused by: liquibase.exception.LiquibaseException: liquibase.exception.MigrationFailedException: Migration failed for change set filepath/example-changeset.sql::example-id::example-author:
		Reason: liquibase.exception.DatabaseException: ORA-00900: invalid SQL statement
In Liquibase 4.15.0 to 4.28.0, Liquibase fixes this by requiring a newline before the / to treat it as a delimiter. Otherwise, the / is treated as a regular character. However, using a newline in the wrong place can lead to the error message "ORA-00922: missing or invalid option" on Oracle databases.

In Liquibase 4.29.0, Liquibase fixes this by requiring a newline after the / to treat it as a delimiter, not before. Otherwise, the / is treated as a regular character.
```
to

`In Liquibase 4.29.2, the / is treated as a regular character when --strict=true and as a delimiter when  --strict=false or not specified.`
is it correct @filipelautert 
https://docs.liquibase.com/change-types/enddelimiter-sql.html

**Comment by filipelautert** [comment]
> `In Liquibase 4.29.2, the / is treated as a regular character when --strict=true and as a delimiter when --strict=false or not specified.`
> is it correct @filipelautert ?
> https://docs.liquibase.com/change-types/enddelimiter-sql.html

@rberezen yes, this seems a good description.

**Comment by adrian-velonis1** [comment]
[DOC TICKET] https://datical.atlassian.net/browse/PD-5023

**Pull request merged**